### PR TITLE
Drop zone: rewrite without provider

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This PR is best reviewed [without whitespace changes](https://github.com/WordPress/gutenberg/pull/30310/files?diff=split&w=1).

Currently, drag and drop is controlled by a provider that keeps track of all the drop zone, checks to see which one is being dragged over and updates state for that drop zone with position info.

This seems all rather complex (and a source of many headaches) and I see no good reason for each drop zone to not manage its own state and events properly.

I've also removed storing position coordinates to state and instead using a callback to signal the position to the drop zone implementor, which can set its own state based on the coordinates. This way, we don't need to continuously set state for position changes but only for data derived from it.

| Before  | After |
| - | - |
| drag-with-provider | drag-without-provider |

## How has this been tested?

Test drag and drop of blocks in the editor, inner blocks, and placeholders.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-editor/src/components/use-block-drop-zone/index.js [hunk 1]
```diff
@@ -120,34 +118,26 @@ export default function useBlockDropZone( {
...
 		isDisabled: isLockedAll,
-		withPosition: true,
 		...dropEventHandlers,
-	} );
-
-	useEffect( () => {
-		if ( position ) {
-			const blockElements = Array.from( element.current.children );
-
+		onDragOver( event ) {
```
Comment (talldan) [inline packages/block-editor/src/components/use-block-drop-zone/index.js]: Much nicer API 👍

File: packages/components/src/drop-zone/provider.js [hunk 2]
```diff
@@ -1,262 +1,11 @@
...
+import deprecated from '@wordpress/deprecated';
 
 export default function DropZoneProvider( { children } ) {
-	return (
-		<DropZoneContextProvider>
-			<DropContainer>{ children }</DropContainer>
-		</DropZoneContextProvider>
-	);
+	deprecated( 'wp.components.DropZoneProvider', {
+		hint: 'wp.component.DropZone no longer needs a provider.',
```
Comment (talldan) [inline packages/components/src/drop-zone/provider.js]: Could be good to add something extra here like "This component is now safe to remove from your code" for a nicer developer experience.

File: packages/block-editor/src/components/block-draggable/index.js [hunk 3]
```diff
@@ -76,6 +76,7 @@ const BlockDraggable = ( {
 		<Draggable
 			cloneClassname={ cloneClassname }
 			elementId={ elementId }
+			__experimentalTransferDataType="wp-blocks"
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-draggable/index.js:79]: This sounds like a duplication of the "type" property in `transferData`

File: packages/block-editor/src/components/block-list/insertion-point.js [hunk 4]
```diff
@@ -233,7 +233,9 @@ function InsertionPointPopover( {
 				tabIndex={ -1 }
 				onClick={ onClick }
 				onFocus={ onFocus }
-				className={ className }
+				className={ classnames( className, {
+					'is-with-inserter': showInsertionPointInserter,
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/insertion-point.js:237]: Why is this needed?

File: packages/block-editor/src/components/block-navigation/use-block-navigation-drop-zone.js [hunk 5]
```diff
@@ -275,45 +189,85 @@ function getBlockNavigationDropTarget( blocksData, position ) {
...
+		target || {};
 
 	const dropEventHandlers = useOnBlockDrop(
 		targetRootClientId,
 		targetBlockIndex
 	);
 
-	const { position, type: dragEventType } = useDropZone( {
+	const ref = useDropZone( {
 		element: ref,
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-navigation/use-block-navigation-drop-zone.js]: Something is weird here, we pass a ref variable and we return it? 

Also, it seems like ideally we'd just remove the element argument right?

File: packages/block-editor/src/components/inner-blocks/index.js [hunk 6]
```diff
@@ -155,7 +155,12 @@ export function useInnerBlocksProps( props = {}, options = {} ) {
 		[ clientId, isSmallScreen ]
 	);
 
-	const ref = useMergeRefs( [ props.ref, fallbackRef ] );
+	const ref = useMergeRefs( [
+		props.ref,
+		useBlockDropZone( {
```
Comment (youknowriad) [inline packages/block-editor/src/components/inner-blocks/index.js:160]: Just to clarify my own understanding, you basically avoided the "wrapperRef" prop by calling this in both inner blocks and the root list right?

File: packages/components/src/drop-zone/index.js [hunk 7]
```diff
@@ -2,67 +2,190 @@
...
+			element.addEventListener( 'drop', onDrop );
+			element.addEventListener( 'dragenter', onDragEnter );
+			element.addEventListener( 'dragover', onDragOver );
+			element.addEventListener( 'dragleave', onDragLeave );
+			ownerDocument.addEventListener( 'mouseup', resetDragState );
+			// Note that `dragend` doesn't fire consistently for file and HTML drag
+			// events where the drag origin is outside the browser window.
+			// In Firefox it may also not fire if the originating node is removed.
+			ownerDocument.addEventListener( 'dragend', resetDragState );
+			ownerDocument.addEventListener( 'dragover', onGlobalDragOver );
```
Comment (youknowriad) [inline packages/components/src/drop-zone/index.js]: Any concern about the number of event listeners this will create? With the provider it was just creating a global one right?

File: packages/edit-post/src/components/layout/index.js [hunk 8]
```diff
@@ -175,7 +170,6 @@ function Layout( { styles } ) {
 		},
 		[ entitiesSavedStatesCallback ]
 	);
-	const ref = useDrop( ref );
```
Comment (youknowriad) [inline packages/edit-post/src/components/layout/index.js:175]: What was this for?

File: packages/components/src/drop-zone/style.scss [hunk 9]
```diff
@@ -61,3 +61,11 @@
...
 	height: 100%;
 }
+
+[data-is-dragging] * {
+	pointer-events: none;
+}
+
+[data-is-drop-target] {
+	pointer-events: all;
+}
```
Comment (youknowriad) [inline packages/components/src/drop-zone/style.scss]: Any concern about these being too global?  should we have a `wp` prefix there or something?

**Review by youknowriad — COMMENTED** [review]
This is very cool. I a have a few questions :)

File: packages/block-editor/src/components/block-draggable/index.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/index.js:79]: It is, and we should remove it there because it's pretty useless. The type and the data are set separately on the `dataTransfer` event instance.

File: packages/block-editor/src/components/block-list/insertion-point.js (hunk 4 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-list/insertion-point.js:237]: Good question. :) It is needed because I removed pointer events on the inserter indicator, which will otherwise "catch" events when it appears between blocks. This worked fine previously because we were tracking `dragover` events globally and we could still determine a position.

Now, the `is-with-inserter` class is added so we can re-enable pointer events for the click area in the dead zone between blocks and the inserter button. This is only necessary when the insertion point includes an inserter button, which it does not in case of e.g. drag and drop.

File: packages/block-editor/src/components/block-navigation/use-block-navigation-drop-zone.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-navigation/use-block-navigation-drop-zone.js]: Ah, good catch. The `element` and `withPosition` arguments can be removed, I simply forgot it.

File: packages/block-editor/src/components/inner-blocks/index.js (hunk 6 above)
Comment (ellatrix) [reply packages/block-editor/src/components/inner-blocks/index.js:160]: That's correct. And in the future, we should be able to reuse `useInnerBlocksProps` in the `BlockList` component to avoid some code duplication.

File: packages/components/src/drop-zone/index.js (hunk 7 above)
Comment (ellatrix) [reply packages/components/src/drop-zone/index.js]: No, it really won't create that much listeners. There will be listeners for each drop zone, which means each block **list** (not block!) and each media placeholder.

File: packages/components/src/drop-zone/style.scss (hunk 9 above)
Comment (ellatrix) [reply packages/components/src/drop-zone/style.scss]: Hm, I see this a bit as a better version of the `is-something` CSS selectors. We also don't prefix all those with `wp-`? Happy to do it though.

File: packages/components/src/drop-zone/index.js (hunk 7 above)
Comment (youknowriad) [reply packages/components/src/drop-zone/index.js]: > There will be listeners for each drop zone, which means each block list (not block!)

Makes me wonder why not just a single drop zone at the top of the block list? What is the thing forcing us to have a dropzone  per container block?

File: packages/edit-post/src/components/layout/index.js (hunk 8 above)
Comment (ellatrix) [reply packages/edit-post/src/components/layout/index.js:175]: Adding a drop zone provider (attaching event listeners) without adding an extra `div`. :) Good riddance.

File: packages/components/src/drop-zone/style.scss (hunk 9 above)
Comment (youknowriad) [reply packages/components/src/drop-zone/style.scss]: I'm fine  with these but the issue is that they don't target any particular id or class. (same issues applies to is-something). In other words, I'd prefer `.components-dropzone[data-is-drop-target]` but I guess we Don't always have an element on which to apply a class?

Not a bit concern by my main thinking is that these names seem generic enough that they could be used by other libraries...

File: packages/components/src/drop-zone/index.js (hunk 7 above)
Comment (ellatrix) [reply packages/components/src/drop-zone/index.js]: We could try that, but I'd suggest trying things one step at a time. We do something similar for the in-between inserter, so I don't see why it couldn't work, though I have a feeling that we'll be much more reliant on DOM queries, which is not great. That's what I don't like about the current in-between inserter `mousemove` handlers, which I'd like to have another look at soon.

It's worth keeping in mind we used to have a drop zone for each block, which was totally crazy. :D

File: packages/components/src/drop-zone/style.scss (hunk 9 above)
Comment (ellatrix) [reply packages/components/src/drop-zone/style.scss]: Yeah. Tbh, I'm hoping to avoid these attributes entirely in a follow up. I can prefix them for now.

File: packages/block-editor/src/components/block-draggable/index.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/index.js:79]: Would you like me to remove the type key inside the data itself? I was never sure why it was there.

File: packages/components/src/drop-zone/style.scss (hunk 9 above)
Comment (ellatrix) [reply packages/components/src/drop-zone/style.scss]: I prefixed it

File: packages/block-editor/src/components/block-navigation/use-block-navigation-drop-zone.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-navigation/use-block-navigation-drop-zone.js]: Removed in cb93fe7.

File: packages/block-editor/src/components/block-draggable/index.js (hunk 3 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-draggable/index.js:79]: if it's possible to remove yes, we already use at least two types there.

Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/index.js:79]: Two types? 🤔 I'll check when I remove it.

Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/index.js:79]: Hm, ok, I see those two subtypes "inserter" and "block". I'd prefer to look at that in a separate PR because it doesn't really seem relevant here. The reason I'm switching the `dataTransfer` [format](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/setData#arguments) to `wp-blocks` is to make it easier to differentiate it from normal text, html and files dragging. Currently we set it as normal text (which isn't right in the first place), so we'd need to parse the text data simply to know what it is.

Comment (youknowriad) [reply packages/block-editor/src/components/block-draggable/index.js:79]: > The reason I'm switching the dataTransfer format to wp-blocks is to make it easier to differentiate it from normal text, html and files dragging. Currently we set it as normal text (which isn't right in the first place), so we'd need to parse the text data simply to know what it is.

This makes me wonder if this type should be more something like `wp` (not blocks but just something to indicate that it's the components drag/drop) and avoid the prop entirely.

File: packages/components/src/drop-zone/index.js (hunk 7 above)
Comment (talldan) [reply packages/components/src/drop-zone/index.js]: > I have a feeling that we'll be much more reliant on DOM queries, which is not great.

Yeah, I think the target detection code would have to check which inner blocks rectangle it's within, so that it can isolate only child block drop targets. Having separate drop zones makes this work automatically.

File: packages/components/src/drop-zone/index.js [hunk 10]
```diff
@@ -2,67 +2,190 @@
...
+			}
+
+			const throttledOnDragOver = throttle( ( event ) => {
+				if ( onDragOverRef.current ) {
+					onDragOverRef.current( event );
+				}
+			}, 200 );
+
+			function onDragOver( event ) {
+				// Only call onDragOver for the innermost hovered drag zones.
```
Comment (talldan) [inline packages/components/src/drop-zone/index.js]: I was wondering how this would be replicated from the previous way it worked 😄

**Review by talldan — COMMENTED** [review]
The API is much nicer, awesome work. 👍 

In testing I noticed dropping into navigation block submenus no longer works, the submenus don't become visible when dragging and hovering over a link block:
https://github.com/WordPress/gutenberg/blob/f98f2d2915953375a88c7ad113bb2595bb07f59a/packages/block-library/src/navigation-link/edit.js#L61-L101

A quick look and I couldn't see why. `onDrop` is the only event that has stopPropagation.

File: packages/components/src/drop-zone/provider.js (hunk 2 above)
Comment (ellatrix) [reply packages/components/src/drop-zone/provider.js]: Sounds good, will add.

File: packages/block-editor/src/components/use-block-drop-zone/index.js (hunk 1 above)
Comment (ellatrix) [reply packages/block-editor/src/components/use-block-drop-zone/index.js]: I'd still like to make some improvements, but I cannot do them without making the PR bigger to rewrite the drop handlers. I'm excited to do a follow-up, but let's see this through piece by piece. 😅

**Comment by ellatrix** [comment]
> A quick look and I couldn't see why. `onDrop` is the only event that has stopPropagation.

I'll have a look. The drop event was previously also stopped from propagating, but I actually don't see why this is necessary if we check `defaultPrevented`.

File: packages/block-editor/src/components/block-draggable/index.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/index.js:79]: Not sure I'm following. What should be `wp`? The `dataTransfer` format?

File: packages/components/src/drop-zone/index.js (hunk 10 above)
Comment (ellatrix) [reply packages/components/src/drop-zone/index.js]: Yeah, this is so much simpler than checking element rectangles. :)

**Comment by ellatrix** [comment]
@talldan I fixed the issue. I had disabled pointer events on children of a drop zone, but we can do it without.

File: packages/block-editor/src/components/block-draggable/index.js (hunk 3 above)
Comment (youknowriad) [reply packages/block-editor/src/components/block-draggable/index.js:79]: Nevermind, it's not that important. I still think we need to dedupe the types but I'm ok with it being a follow-up.

Comment (ellatrix) [reply packages/block-editor/src/components/block-draggable/index.js:79]: Ok, sounds good! 🙂

**Review by talldan — APPROVED** [review]
Sorry it took so long for an approval. I tested again and this is working well 👍 

Just some conflicts to sort out, but hopefully that's straightforward.

**Comment by ellatrix** [comment]
Thanks a lot for reviewing everyone!
